File: LongestSubstringWithoutRepeatingCharacters/sliding_window.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        chars = set()
        longest = 0

        for r in range(len(s)):
            while s[r] in chars:
                chars.remove(s[l])
                l += 1
            chars.add(s[r])
            longest = max(longest, len(chars))

        return longest

File: LongestSubstringWithoutRepeatingCharacters/sliding_window.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        var l = 0;
        var chars = new HashSet<Character>();
        var longest = 0;

        for (var r = 0; r < s.length(); r++) {
            while (chars.contains(s.charAt(r))) {
                chars.remove(s.charAt(l));
                l++;
            }
            chars.add(s.charAt(r));
            longest = Math.max(longest, chars.size());
        }

        return longest;
    }
}
